### PR TITLE
zson writer

### PR DESCRIPTION
**Pull request by mccanne**
These changes allow for reading and writing of zson.  The zeek
reader was refactored into a zson reader that uses a zeek parser
for legacy zeek logs.  This commit also fixes a bug where we
weren't properly handling semicolon termination.  The zeek parsing
tests were broken out into the legacy parsing tests and tests
in zsio for the new zson syntax.

### CODE REVIEW

File: pkg/zsio/zsio_test.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	"strings"
+	"testing"
+
+	"github.com/mccanne/zq/pkg/zson/resolver"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func assertError(t *testing.T, err error, pattern, what string) {
+	assert.NotNilf(t, err, "Received error for %s", what)
```
Comment (nwt) [inline pkg/zsio/zsio_test.go]: Nit: Use assert.Error.

File: pkg/zsio/writer.go [hunk 2]
```diff
@@ -1,20 +1,102 @@
...
+}
+
+func zsonEscape(s string) string {
+	if s == "-" {
+		return "\\-"
+	}
+
+	s = strings.ReplaceAll(s, "\\", "\\\\")
+	s = strings.ReplaceAll(s, ";", "\\;")
+	s = strings.ReplaceAll(s, "\n", "\\n")
```
Comment (nwt) [inline pkg/zsio/writer.go]: `\-` and `\;` are not valid JavaScript escape sequences. I vote for replacing them with `\xdd` sequences.

File: pkg/zsio/reader.go [hunk 3]
```diff
@@ -1,21 +1,223 @@
...
+	return &Reader{
+		scanner:     skim.NewScanner(reader, buffer, MaxLineSize),
+		zeek:        zeek.NewParser(r),
+		descriptors: make(map[int]*zson.Descriptor),
+		resolver:    r,
+	}
+}
+
+// getline returns the next line skipping blank lines and too-long lines
+// XXX for zq, we should probabluy return line-too-long error
```
Comment (mattnibs) [inline pkg/zsio/reader.go]:
```suggestion
// XXX for zq, we should probably return line-too-long error
```

**Comment by mattnibs** [comment]
Maybe add a system test or two :) ?

**Comment by philrz** [comment]
I was doing some black box testing with this branch and noticed a couple things:

1. While `zq -f zson "*" *.log` worked and showed me ZSON format for all my data in one output stream (yay!), it looks like we need to make reference to it in the help text:

```
    -f format for output data [text,table,zeek,json,ndjson,raw] (default "table")
```

2. Further along those lines, is now the time to change that default output format from `table` to `zson`? I'd be in favor of that. I like the `table` format for quick readability, but as a default, it feels way too hazardous. I don't like the idea of a customer's first query being a grump that shows no data:

```
# zq -f table "*" conn.log 
output table too big
```

We've been through rounds of debates on `table` before and I don't want to re-litigate that. I'd be happier to just embrace the new goodness of ZSON and let `table` hang out as the non-default that it is.

File: pkg/zsio/reader.go (hunk 3 above)
Comment (mccanne) [reply pkg/zsio/reader.go]: +1

File: pkg/zsio/writer.go (hunk 2 above)
Comment (mccanne) [reply pkg/zsio/writer.go]: ok, let's do this in a subsequent PR along with the change to the spec.  I think it's worthing reworking the logic here anyway.

File: pkg/zsio/zsio_test.go (hunk 1 above)
Comment (mccanne) [reply pkg/zsio/zsio_test.go]: +1

File: pkg/zsio/writer.go (hunk 2 above)
Comment (nwt) [reply pkg/zsio/writer.go]: Works for me.

**Pull request merged**